### PR TITLE
Refactor metadata validation

### PR DESCRIPTION
**Pull request by iamyulong**
## Summary

Refactor metadata validation, particularly checking length before validating URLs.

### CODE REVIEW

File: radix-engine-tests/assets/blueprints/metadata2/src/lib.rs
Comment (dhedey) [inline radix-engine-tests/assets/blueprints/metadata2/src/lib.rs:13]: Is this commented-out code intended to be deleted? :)

Comment (dhedey) [reply radix-engine-tests/assets/blueprints/metadata2/src/lib.rs:13]: Ahh no it's explaining how the below was generated, my bad! All good.

Perhaps we could have a comment here briefly explaining that? But only if you feel like it, no worries, I was just being slow.

File: radix-engine-tests/benches/metadata_validation.rs
Comment (dhedey) [inline radix-engine-tests/benches/metadata_validation.rs:13]: MINOR: This is always, valid, right? So we should probably just `unwrap()` here - otherwise greedy allocation of a string might slow down the test a little.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (dhedey) [inline radix-engine/src/object_modules/metadata/package.rs:20]: This will require temporarily using `, regenerate` to update the `VersionedLocalTransactionExecution`.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (dhedey) [inline radix-engine/src/object_modules/metadata/package.rs:19]: We can't remove these error variants as it breaks the error deserialization of historic receipts in the node (until we fix that to store historic schemas - which I really want to do one day soon ;))

Even if they're unused, we need to keep them for now I'm afraid.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (dhedey) [inline radix-engine/src/object_modules/metadata/package.rs:266]: How confident are we that these refactors are 100% backwards-compatible across Cuttlefish in terms of protocol-visible changes? e.g. we're changing which errors we return (which I think is safe I think assuming we only error in the same cases). And e.g. we aren't making any sys-calls/billing in the middle, which seems safe?

I see we've replaced some encode unwraps with e.g. `MetadataKeyValidationError::InvalidValue` - I guess we don't actually assume this to be hit, and indeed if it unwrapped previously, it shouldn't introduce any issues?

File: radix-engine/src/object_modules/metadata/package.rs
Comment (dhedey) [inline radix-engine/src/object_modules/metadata/package.rs:264]: Can we perhaps make it not `pub`?

Perhaps we should make it not `pub` and rename it to `init_system_struct_from_pre_validated_data` or something?

And then we can re-expose some higher-level `pub` method which validates and then inits?

**Review by dhedey — APPROVED** [review]
Nice! A couple of questions I'd like discussing before merge, but looks good :)

File: radix-engine-tests/assets/blueprints/metadata2/src/lib.rs
Comment (iamyulong) [reply radix-engine-tests/assets/blueprints/metadata2/src/lib.rs:13]: Will some add notes.

File: radix-engine-tests/benches/metadata_validation.rs
Comment (iamyulong) [reply radix-engine-tests/benches/metadata_validation.rs:13]: Happy to unwrap.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (iamyulong) [reply radix-engine/src/object_modules/metadata/package.rs:20]: Will do.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (iamyulong) [reply radix-engine/src/object_modules/metadata/package.rs:19]: You're right. Thought we had that error to string conversion already.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (iamyulong) [reply radix-engine/src/object_modules/metadata/package.rs:266]: > How confident are we that these refactors are 100% backwards-compatible across Cuttlefish in terms of protocol-visible changes

I did think about this thoroughly. As far I can tell, there is only error code difference which should not affect consensus, but would appreciate another pair of eyes.

> I see we've replaced some encode unwraps with e.g. MetadataKeyValidationError::InvalidValue - I guess we don't actually assume this to be hit, and indeed if it unwrapped previously, it shouldn't introduce any issues?

This is right, considering `MetadataInit` often comes from an sbor-encoded inputs.

Keeping it in the validation logic so it's self-contained and no need to make assumption about sbor-able.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (iamyulong) [reply radix-engine/src/object_modules/metadata/package.rs:264]: Will make it non-public.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (dhedey) [reply radix-engine/src/object_modules/metadata/package.rs:19]: We do have error to string conversion, but to save space, it happens on the API at response serving time; and uses the _current_ `RuntimeError`. I'd like to store some historic schemas at each protocol update in the node so we can convert with these old schemas whilst still updating the errors.

File: radix-engine/src/object_modules/metadata/package.rs
Comment (dhedey) [reply radix-engine/src/object_modules/metadata/package.rs:268]: Sounds great. And yes, I agree that I think it's safe, so I'm happy with this.

**Pull request merged**
